README
#############################################################################################################
# EULA:
#
# WIVE FIRMWARE IS FREE FOR THE NON-COMMERCIAL USE ONLY.
#
# Conditions of commercial use non GPL (or other not viral license) components are discussed individually.
# For a more detailed discussion please mail me.
# News and changes in Russian: http://wive-ng.sf.net
# Bug report please send to sfstudio[at]mail.ru
#############################################################################################################

BUILD:
1) cd toolchain and start build_toolchain.sh as root
2) cd in projectroot (/opt/Wive-MT) and start compile <target name> or compile-releases for build all targets

#############################################################################################################


History
<hr>
<b>WIVE FIRMWARE IS FREE FOR THE NON-COMMERCIAL USE ONLY.</b>
<br>
<b>Conditions of commercial use non GPL (or other not viral license) components are discussed individually.</b>
<hr>
<b>News and changes in Russian:</b><br>
<a href="http://wive-ng.sf.net">wive-ng.sf.net - development opensource routers firmware</a><br>
<br>
<b>Bug report please send sfstudio[at]mail.ru</b>
<hr>
<b>Changes:</b><br>
<b>1.0.x</b><br>
1) Fix configs scripts for support bash >=4 and change product name.<br>
2) Kernel version up to 2.6.21.7.<br>
3) Redesign filesystem. Busybox up to 1.15.3.<br>
4) PPPD up to 2.4.5 with more patches as acell-ptpp/l2tp/gigaworld/allow-mppe/etc.<br>
5) Add ACELL-PPTP Kernel support.<br>
6) Mount ramfs particions in kernel mode.<br>
7) BRIDGE: Pause frames shouldn't be passed up by driver anyway.<br>
8) Add ESFQ support.<br>
9) Initial auto detect memory size (fix me 32bit mode support).<br>
10) Add udpxy.<br>
11) Code cleanup.<br>
12) Remove ASUS Code.<br>
13) Original Ralink gui now work.<br>
14) pptp/l2tp/pppoe tun now works.<br>
15) Full rc scrips is write in rwfs.<br>
16) RT3052F embed switch full work now.<br>
17) Rewrite pptp/l2tp/pppoe scripts to support buggy nas (corbina etc).<br>
18) miniupnpd now work too.<br>
19) Add rwfs support (fs save | fs restore | fs load).<br>
20) Auto set mss size.<br>
21) Masqrade now work correctly.<br>
22) IMQ support added.<br>
23) Initial shaper scripts (default priomap and esfq for decrase torrents prio).<br>
24) Kernel ip account support added.<br>
25) Fix crash on sta list.<br>
26) Update userspace daemons and utility.<br>
27) Some busybox fixes.<br>
28) Auto generate UUID for UPNP service on first upnpd start.<br>
29) Check if ppp is up do not change ip from dhcp.<br>
30) Regenerate mac if drop to defaults.<br>
31) Some services start only then start goahead.<br>
32) Replace some code from original Ralink SDK version 3.3.<br>
33) Integrate new web gui in current sources tree.<br>
34) Static routes get from dhcp server (options 79).<br>
35) busybox update to 1.16.0 version.<br>
36) Replace ntpclient to ntpd from new busybox.<br>
37) Add multiple default gateway support for load balacing.<br>
38) Fix boot on 4/16 devices.<br>
39) RootFS on flash work OK.<br>
40) Fix resolve server ip for pptp/l2tp.<br>
41) Add support others type falsh.<br>
42) Job control fix.<br>
43) Fix some daemons crash with error build in future.<br>
44) fs save now save current date and time to rwfs.<br>
45) Fix configure esw mode and more modes add.<br>
46) Fix pppoe reconnect.<br>
47) Fix pppoe long discovery.<br>
48) Restore default gateway after VPN down.<br>
49) WPS wscd daemon now work.<br>
50) Fix dhcp lease parse.<br>
51) busybox update to 1.16.1 version.<br>
52) Fix pptp/pppoe/l2tp redials.<br>
53) Radvd up to 1.6 version.<br>
54) Masqrade for vpn move to ip-up/ip-down scripts.<br>
55) NAT to ISP local net parallel then vpn work now.<br>
56) Never increase MSS, even when setting it.<br>
57) Backport small bridge patch.<br>
58) Backport sock memalloc skb size from 2.6.23.4 kernel.<br>
59) ACCEL-PPTP add module param description and mtu parametr.<br>
60) Patch to fix cut end symbol at hostname in udhcpd lease.<br>
61) Backport jhash3 patch from upstream.<br>
62) More ipv6 fixes.<br>
63) Improve pppoe perf.<br>
64) pptp module patches from wl500 project.<br>
65) More user/password length to pptp/l2tp mode.<br>
66) More clean init.d scripts.<br>
67) Fix kernel crash on recive multicast.<br>
68) Autodetect RAM size (16-64) work now for all devices.<br>
69) UDPXY work now.<br>
70) Igmpproxy work now.<br>
71) Use TMPFS instead RAMFS. Save memory and increase speed.<br>
72) Fix restore nvram after flash at new device.<br>
73) LZMA crash fix.<br>
74) Mtdblock crash and memory leak fix.<br>
75) Remove cifs. User ssh scp instead.<br>
76) Update wifi driver to 2.3.0.0 version from Ralink.<br>
77) More fix build enviroment.<br>
78) Bridge code fixup and more fast.<br>
<br>
<b>1.1.x</b><br>
1) New WebUI based on goahead.<br>
2) New wifi driver, correct work with broadcom wifi based adapters, WMM, etc...<br>
3) Set of corrections in a kernel (netfiler/routing/ipv6).<br>
4) Rewrite logic in init.d for speedup.<br>
5) Add simple software conntrack fastpath.<br>
6) Cleannat module support for clear conntrack table.<br>
7) More connrack helpers and protocol support.<br>
8) Kernel backport fixes from stable tree.<br>
9) Busybox update to 1.17.x tree.<br>
10) Some scrits rewrite for speedup.<br>
11) Remove hldc and vj support for pppd.<br>
12) ppp connection optimization in kernel.<br>
13) DMZ must works.<br>
14) Sync from 3.4.0.1 RalinkSDK from trandnet TEW-691GR GPL code.<br>
15) libnvram fix crash on reliad settings and some leaks.<br>
16) Fix freeze in 802.11n mode change in EthernetConverterMode.<br>
17) ppp use metric 10 for dgw replace.<br>
18) add more workaround for buggy nas.<br>
19) Portforward work now.<br>
20) MAC/IP/PORT filterd work now.<br>
21) pptp/l2tp now work with new corbina bras.<br>
22) Get staticroutes and msstatic routes from dhcp server work now.<br>
23) WebUI Add watchdog enable/disable function.<br>
24) Fix reboot at apply change in ethernet converter mode.<br>
25) Fix lldt control in WebUI.<br>
26) Add interface select for macip filter and port forward in WebUI.<br>
27) Add 802.11g/n mixed mode and set default.<br>
28) Add samba for smb and wins services.<br>
29) portforward and ipport filter scripts now reboot at start.<br>
30) Content filters based on webstr now works.<br>
31) Use WLAN_MAC_ADDR, WAN_MAC_ADDR, LAN_MAC_ADDR variable in nvram for set macs to interfaces.<br>
32) STA profiles now works.<br>
33) Add netmask for ip filter pages.<br>
34) Add asus rt-g32 full support.<br>
35) Fix part table in SPI and nand flash types.<br>
36) pppd up to current git src 24 Aug 2010.<br>
37) xl2tpd up to 1.2.7 version.<br>
38) Fix conntrack flush.<br>
39) Wireless pages split and optimize.<br>
40) VPN page add auth type select.<br>

user/rc/ppp.c
/*
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation; either version 2 of
 * the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston,
 * MA 02111-1307 USA
 */
/*
 * ppp scripts
 *
 * Copyright 2004, Broadcom Corporation
 * All Rights Reserved.
 * 
 * THIS SOFTWARE IS OFFERED "AS IS", AND BROADCOM GRANTS NO WARRANTIES OF ANY
 * KIND, EXPRESS OR IMPLIED, BY STATUTE, COMMUNICATION OR OTHERWISE. BROADCOM
 * SPECIFICALLY DISCLAIMS ANY IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS
 * FOR A SPECIFIC PURPOSE OR NONINFRINGEMENT CONCERNING THIS SOFTWARE.
 *
 * $Id: ppp.c,v 1.1.1.1 2007/01/25 12:52:21 jiahao_jhou Exp $
 */

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <net/route.h>
#include <sys/ioctl.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <errno.h>
#include <ctype.h>

#include <nvram/bcmnvram.h>
#include <netconf.h>
#include <shutils.h>
#include <rc.h>

/*
* parse ifname to retrieve unit #
*/
int
ppp_ifunit(char *ifname)
{
	if (strncmp(ifname, "ppp", 3))
		return -1;
	if (!isdigit(ifname[3]))
		return -1;
	return atoi(&ifname[3]);
}

/*
 * Called when link comes up
 */
int
ipup_main(int argc, char **argv)
{
	FILE *fp;
	char *wan_ifname = safe_getenv("IFNAME");
	char *value;
	char buf[256];
	int unit;
	char tmp[100], prefix[] = "wanXXXXXXXXXX_";

	umask(022);
 
	if ((unit = ppp_ifunit(wan_ifname)) < 0)
		return -1;
	
	snprintf(prefix, sizeof(prefix), "wan%d_", unit);

	/* Touch connection file */
	if (!(fp = fopen(strcat_r("/tmp/ppp/link.", wan_ifname, tmp), "a"))) {
		perror(tmp);
		return errno;
	}
	fclose(fp);

	if (!nvram_get(strcat_r(prefix, "ifname", tmp)))
		return -1;

	if ((value = getenv("IPLOCAL"))) {
		ifconfig(wan_ifname, IFUP,
			 value, "255.255.255.255");
		nvram_set(strcat_r(prefix, "ipaddr", tmp), value);
		nvram_set(strcat_r(prefix, "netmask", tmp), "255.255.255.255");
	}

        if ((value = getenv("IPREMOTE")))
		nvram_set(strcat_r(prefix, "gateway", tmp), value);
	strcpy(buf, "");
	if (getenv("DNS1"))
		sprintf(buf, "%s", getenv("DNS1"));
	if (getenv("DNS2"))
		sprintf(buf + strlen(buf), "%s%s", strlen(buf) ? " " : "", getenv("DNS2"));
	nvram_set(strcat_r(prefix, "dns", tmp), buf);

	wan_up(wan_ifname);

	logmessage(nvram_safe_get("wan_proto_t"), "connect to ISP");
	wanmessage("");

	return 0;
}

/*
 * Called when link goes down
 */
int
ipdown_main(int argc, char **argv)
{
	char *wan_ifname = safe_getenv("IFNAME");
	int unit;
	char tmp[100], prefix[] = "wanXXXXXXXXXX_";
	
	umask(022);

	if ((unit = ppp_ifunit(wan_ifname)) < 0)
		return -1;
	
	snprintf(prefix, sizeof(prefix), "wan%d_", unit);

	if (!nvram_get(strcat_r(prefix, "ifname", tmp)))
		return -1;

	wan_down(wan_ifname);

	unlink(strcat_r("/tmp/ppp/link.", wan_ifname, tmp));

	preset_wan_routes(wan_ifname);

	logmessage(nvram_safe_get("wan_proto_t"), "Disconnected");
	wanmessage(pppstatus(tmp));

	return 0;
}

